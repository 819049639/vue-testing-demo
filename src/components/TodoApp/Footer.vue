<!--
 * @Author: wangqiang@feewee.cn
 * @Date: 2022-07-19 14:46:43
 * @LastEditors: wangqiang@feewee.cn
 * @LastEditTime: 2022-07-20 09:14:14
-->
<template>
  <footer class="footer">
    <!-- This should be `0 items left` by default -->
    <span class="todo-count"
      ><strong data-testid="done-todos-count">{{ doneTodosCount }}</strong> item
      left</span
    >
    <!-- Remove this if you don't implement routing -->
    <ul class="filters">
      <li>
        <!-- <a :class="{ selected: $route.path === '/' }" href="#/">All</a> -->
        <router-link exact to="/" >All</router-link>
      </li>
      <li>
        <!-- <a :class="{ selected: $route.path === '/active' }" href="#/active">Active</a> -->
        <router-link to="/active" >Active</router-link>
      </li>
      <li>
        <!-- <a
          :class="{ selected: $route.path === '/completed' }"
          href="#/completed"
          >Completed</a
        > -->
        <router-link to="/completed" >Completed</router-link>
      </li>
    </ul>
    <!-- Hidden if no completed items are left â†“ -->
    <button
      v-if="isClearCompletedShow"
      data-testid="clear-completed"
      class="clear-completed"
      @click="$emit('clear-completed')"
    >
      Clear completed
    </button>
  </footer>
</template>
<script>
export default {
  name: 'TodoAppFooter',
  props: {
    todos: {
      type: Array,
      required: true
    }
  },
  computed: {
    doneTodosCount () {
      return this.todos.filter((t) => !t.done).length
    },
    isClearCompletedShow () {
      return this.todos.find((t) => t.done)
    }
  }
}
</script>
