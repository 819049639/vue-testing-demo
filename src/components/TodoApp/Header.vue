<!--
 * @Author: wangqiang@feewee.cn
 * @Date: 2022-07-19 14:44:04
 * @LastEditors: wangqiang@feewee.cn
 * @LastEditTime: 2022-07-19 14:59:21
-->
<template>
  <header class="header">
    <h1>todos</h1>
    <input
      class="new-todo"
      data-testid="new-todo"
      placeholder="What needs to be done?"
      autofocus
      @keyup.enter="handleNewTodo"
    />
  </header>
</template>
<script>
export default {
  name: 'TodoAppHeader',
  methods: {
    handleNewTodo (e) {
      const value = e.target.value.trim()
      if (!value) {
        return
      }
      this.$emit('new-todo', value)
      e.target.value = ''
    }
  }
}
</script>
