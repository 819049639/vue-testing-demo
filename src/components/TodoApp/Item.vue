<!--
 * @Author: wangqiang@feewee.cn
 * @Date: 2022-07-19 14:47:45
 * @LastEditors: wangqiang@feewee.cn
 * @LastEditTime: 2022-07-19 16:53:32
-->
<template>
  <li
    data-testid="todo-item"
    :class="{ completed: todo.done, editing: isEditing }"
  >
    <div class="view">
      <input
        :checked="todo.done"
        @click="
          $emit('change_done', {
            id: todo.id,
            done: $event.target.checked,
          })
        "
        data-testid="todo-done"
        class="toggle"
        type="checkbox"
      />
      <label data-testid="todo-text" @dblclick="isEditing = true">{{
        todo.text
      }}</label>
      <button
        data-testid="todo-delete"
        class="destroy"
        @click="$emit('delete-todo', todo.id)"
      ></button>
    </div>
    <input
      v-focus="isEditing"
      data-testid="todo-edit"
      class="edit"
      :value="todo.text"
      @blur="isEditing = false"
      @keyup.enter="editTodo"
      @keyup.esc="isEditing = false"
    />
  </li>
</template>
<script>
export default {
  name: 'TodoAppItem',
  props: {
    todo: {
      type: Object,
      required: true
    }
  },
  directives: {
    focus (element, binding) {
      if (binding.value) {
        element.focus()
      }
    }
  },
  data () {
    return {
      isEditing: false
    }
  },
  methods: {
    editTodo (e) {
      this.$emit('edit-todo', { id: this.todo.id, text: e.target.value.trim() })
      this.isEditing = false
    }
  }
}
</script>
