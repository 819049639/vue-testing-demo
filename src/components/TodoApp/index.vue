<!--
 * @Author: wangqiang@feewee.cn
 * @Date: 2022-07-19 14:35:25
 * @LastEditors: wangqiang@feewee.cn
 * @LastEditTime: 2022-07-19 18:01:41
-->
<template>
  <section class="todoapp">
    <Header @new-todo="handleNewTodo" />
    <!-- This section should be hidden by default and shown when there are todos -->
    <section class="main">
      <input
        data-testid="toggle-all"
        id="toggle-all"
        class="toggle-all"
        type="checkbox"
        v-model="toggleAll"
      />
      <label for="toggle-all">Mark all as complete</label>
      <ul class="todo-list">
        <!-- These are here just to show the structure of the list items -->
        <!-- List items should get the class `editing` when editing and `completed` when marked as completed -->
        <!-- <li class="completed">
          <div class="view">
            <input class="toggle" type="checkbox" checked />
            <label>Taste JavaScript</label>
            <button class="destroy"></button>
          </div>
          <input class="edit" value="Create a TodoMVC template" />
        </li> -->
        <Item
          v-for="todo in filterTodos"
          :key="todo.id"
          :todo="todo"
          @change_done="changeDone"
          @delete-todo="deleteTodo"
          @edit-todo="editTodo"
        />
      </ul>
    </section>
    <!-- This footer should be hidden by default and shown when there are todos -->
    <Footer :todos="todos" @clear-completed="clearCompleted" />
  </section>
</template>
<script>
import Header from './Header.vue'
import Footer from './Footer.vue'
import Item from './Item.vue'

export default {
  name: 'TodoApp',
  components: {
    Header,
    Footer,
    Item
  },
  data () {
    return {
      todos: [
        { id: 1, text: 'eat', done: false },
        { id: 2, text: 'play', done: true },
        { id: 3, text: 'sleep', done: false }
      ]
    }
  },
  computed: {
    toggleAll: {
      get () {
        // 设置 toggleAll 的选中状态
        return this.todos.length && this.todos.every((todo) => todo.done)
      },
      set (checked) {
        this.todos.forEach((todo) => {
          todo.done = checked
        })
      }
    },
    filterTodos () {
      // 获取路由路径
      const path = this.$route.path
      // 根据路由路径过滤数据
      // / 返回所有数据
      // /active 所有未完成任务
      // /completed 所有完成任务
      switch (path) {
        case '/':
          return this.todos
        case '/active':
          return this.todos.filter((t) => !t.done)
        case '/completed':
          return this.todos.filter((t) => t.done)
        default:
          return this.todos
      }
    }
  },
  methods: {
    handleNewTodo (text) {
      const lastTodo = this.todos[this.todos.length - 1]
      this.todos.push({
        id: lastTodo ? lastTodo.id + 1 : 1,
        text,
        done: false
      })
    },
    changeDone (data) {
      this.todos = this.todos.map((todo) => ({
        ...todo,
        done: data.id === todo.id ? data.done : todo.done
      }))
    },
    deleteTodo (id) {
      this.todos = this.todos.filter((todo) => todo.id !== id)
    },
    editTodo (data) {
      if (!data.text) {
        const index = this.todos.findIndex((todo) => todo.id === data.id)
        if (index !== -1) {
          this.todos.splice(index, 1)
        }
      } else {
        this.todos = this.todos.map((todo) => ({
          ...todo,
          text: data.id === todo.id ? data.text : todo.text
        }))
      }
    },
    clearCompleted () {
      this.todos = this.todos.filter((todo) => !todo.done)
    }
  }
}
</script>
