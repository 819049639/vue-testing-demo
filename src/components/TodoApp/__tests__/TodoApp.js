/*
 * @Author: wangqiang@feewee.cn
 * @Date: 2022-07-19 15:04:01
 * @LastEditors: wangqiang@feewee.cn
 * @LastEditTime: 2022-07-20 09:29:27
 */
import { shallowMount } from '@vue/test-utils'
import Vue from 'vue'
import TodoApp from '../index.vue'
import Item from '../Item.vue'

describe('TodoApp.vue', () => {
  /** @type {import('@vue/test-utils').Wrapper} */
  let wrapper = null

  beforeEach(async () => {
    const todos = [
      { id: 1, text: 'eat', done: false },
      { id: 2, text: 'play', done: true },
      { id: 3, text: 'sleep', done: false }
    ]
    const $route = {
      path: '/'
    }
    wrapper = shallowMount(TodoApp, {
      mocks: {
        $route
      }
    })
    await wrapper.setData({
      todos
    })
  })
  test('New Todo', async () => {
    const text = 'play'
    wrapper.vm.handleNewTodo(text)
    const todo = wrapper.vm.todos.find((t) => t.text === text)
    expect(todo).toBeTruthy()
  })

  test('Todo List', async () => {
    expect(wrapper.findAllComponents(Item).length).toBe(
      wrapper.vm.todos.length
    )
  })

  test('Todo Delete', async () => {
    await wrapper.vm.deleteTodo(1)
    expect(wrapper.vm.todos.length).toBe(2)
    expect(wrapper.findAllComponents(Item).length).toBe(2)
  })

  test('Todo Delete', async () => {
    await wrapper.vm.deleteTodo(13)
    expect(wrapper.vm.todos.length).toBe(3)
    expect(wrapper.findAllComponents(Item).length).toBe(3)
  })

  test('Todo Edit', async () => {
    const todo = { id: 1, text: 'hello' }
    await wrapper.vm.editTodo(todo)
    expect(wrapper.vm.todos[0].text).toBe(todo.text)

    todo.text = ''
    await wrapper.vm.editTodo(todo)
    expect(wrapper.vm.todos.find((t) => t.id === todo.id)).toBeFalsy()
  })

  test('toggle all', async () => {
    const toggleAll = wrapper.find('input[data-testid="toggle-all"]')
    await toggleAll.setChecked(true)
    // 断言所有的子任务都被选中
    wrapper.vm.todos.forEach((todo) => {
      expect(todo.done).toBeTruthy()
    })

    // 取消全选状态
    await toggleAll.setChecked(false)
    // 断言所有的子任务都未被选中
    wrapper.vm.todos.forEach((todo) => {
      expect(todo.done).toBeFalsy()
    })
  })

  test('toggle all state', async () => {
    const toggleAll = wrapper.find('input[data-testid="toggle-all"]')
    // 让所有任务都变成完成状态
    wrapper.vm.todos.forEach((todo) => {
      todo.done = true
    })
    await Vue.nextTick()
    // 断言 toggleAll 也被选中了
    expect(toggleAll.element.checked).toBeTruthy()

    // 取消某个任务未完成，断言 toggleAll 未选中
    wrapper.vm.todos[0].done = false
    await Vue.nextTick()
    expect(toggleAll.element.checked).toBeFalsy()
  })

  test('clear all completed', async () => {
    wrapper.vm.clearCompleted()
    await Vue.nextTick()
    expect(wrapper.vm.todos).toEqual([
      { id: 1, text: 'eat', done: false },
      { id: 3, text: 'sleep', done: false }
    ])
  })

  test('Filter Todos', async () => {
    // 将路由导航到 /
    wrapper.vm.$route.path = '/'
    await Vue.nextTick()
    // 断言 filterTodos 是全部任务列表
    expect(wrapper.vm.filterTodos).toEqual([
      { id: 1, text: 'eat', done: false },
      { id: 2, text: 'play', done: true },
      { id: 3, text: 'sleep', done: false }
    ])

    // 将路由导航到 /active
    wrapper.vm.$route.path = '/active'
    await Vue.nextTick()
    // 断言 filterTodos 是未完成列表
    expect(wrapper.vm.filterTodos).toEqual([
      { id: 1, text: 'eat', done: false },
      { id: 3, text: 'sleep', done: false }
    ])

    // 将路由导航到 /completed
    wrapper.vm.$route.path = '/completed'
    await Vue.nextTick()
    // 断言 filterTodos 是已完成列表
    expect(wrapper.vm.filterTodos).toEqual([
      { id: 2, text: 'play', done: true }
    ])
  })

  test('snapshot', async () => {
    expect(wrapper.html()).toMatchSnapshot()
  })
})
