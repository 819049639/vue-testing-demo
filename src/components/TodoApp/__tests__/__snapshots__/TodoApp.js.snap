// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoApp.vue snapshot 1`] = `
<section class="todoapp">
  <header-stub></header-stub>
  <section class="main"><input data-testid="toggle-all" id="toggle-all" type="checkbox" class="toggle-all"> <label for="toggle-all">Mark all as complete</label>
    <ul class="todo-list">
      <item-stub todo="[object Object]"></item-stub>
      <item-stub todo="[object Object]"></item-stub>
      <item-stub todo="[object Object]"></item-stub>
    </ul>
  </section>
  <footer-stub todos="[object Object],[object Object],[object Object]"></footer-stub>
</section>
`;
